### PR TITLE
try resolve promise type if not have Promise symbole

### LINKED ISSUE #22782: transpileModule doesn't emit correct metadata of return type for async function
I got a problem when using `transpileModule` with `emitDecoratorMetadata` flag on, the result is inconsistent with what I got from command line tsc.

**TypeScript Version:**  2.8.0-dev.20180321

**Search Terms:** 
experimentalDecorators transpileModule

**Code**
```js
'use strict';

const tsc = require("typescript");
const result = tsc.transpileModule(`
class Foo {
    @AnyDecorator()
    public async foobar(): Promise<void> {

    }
}
`, {
    compilerOptions: {
        emitDecoratorMetadata:true,
        experimentalDecorators:true,
        lib: ["lib.es2015.d.ts"]
    }
});

console.log(result.outputText);
```

**Expected behavior:**
This should decorate the function foobar with the return type `Promise`, like:
```js
__decorate([
    AnyDecorator(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Foo.prototype, "foobar");
```

**Actual behavior:**
```js
__decorate([
    AnyDecorator(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], Foo.prototype, "foobar");
```

However, if I remove the return type from foobar's signature:
```js
'use strict';

const tsc = require("typescript");
const result = tsc.transpileModule(`
class Foo {
    @AnyDecorator()
    public async foobar() {

    }
}
`, {
    compilerOptions: {
        emitDecoratorMetadata:true,
        experimentalDecorators:true,
        lib: ["lib.es2015.d.ts"]
    }
});

console.log(result.outputText);
```
everything is correct, and return type is now `Promise`.

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #22782

### CODE REVIEW

**Comment by mhegazy** [comment]
@rbuckton can you please take a look at this change...

**Comment by weswigham** [comment]
Aside: If we're improving metadata emit, should we reconsider https://github.com/Microsoft/TypeScript/pull/18999?

Of substance: Rather than this change, I believe in `serializeReturnTypeOfNode` we should probably just put the `isAsyncFunction` condition before the `isFunctionLike(node) && node.type` condition, since all async functions should return a `Promise` regardless of what their return type is. I don't really know if all the conditionals suggested to be placed in the emit are really inline with that we emit right now, either - but if we want that we should probably update it for all `Promise` serialized types, and not just ones present via `.type`.

File: src/compiler/transformers/ts.ts [hunk 1]
```diff
@@ -1991,6 +1991,17 @@ namespace ts {
...
+                    if (isIdentifier(node.typeName) && node.typeName.escapedText === "Promise") {
+                        return createLogicalOr(
+                            createLogicalAnd(
+                                createTypeCheck(createIdentifier("Promise"), "function"),
+                                createIdentifier("Promise")
+                            ),
+                            createIdentifier("Object")
+                        );
+                    }
+                    // make 'no-switch-case-fall-through' happy
```
Comment (alfaproject) [inline src/compiler/transformers/ts.ts]: You can disable that rule inline if you are adding a comment anyway

File: src/compiler/transformers/ts.ts [hunk 2]
```diff
@@ -1991,6 +1991,8 @@ namespace ts {
                     return createIdentifier("Promise");
 
                 case TypeReferenceSerializationKind.ObjectType:
+                    return createIdentifier("Object");
```
Comment (alfaproject) [inline src/compiler/transformers/ts.ts]: This is already the default, so why have this explicitly? o:

File: src/compiler/transformers/ts.ts [hunk 3]
```diff
@@ -1794,12 +1794,12 @@ namespace ts {
          * @param node The node that should have its return type serialized.
          */
         function serializeReturnTypeOfNode(node: Node): SerializedTypeNode {
-            if (isFunctionLike(node) && node.type) {
-                return serializeTypeNode(node.type);
-            }
-            else if (isAsyncFunction(node)) {
+            if (isAsyncFunction(node)) {
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: This isn't correct if you are targeting ES5 with async functions and a third-party `Promise` implementation:

```ts
import { SomePromise } from "some-package";
class C {
  @dec
  async function foo(): SomePromise {
  }
}
```

In that case, we want to emit `SomePromise`, not `Promise`.

You probably need to add an additional check that the async function either:

- Has no return type
- Has a return type that cannot be resolved as a value and we are in `isolatedModules`.

**Pull request closed without merging**